### PR TITLE
T07: Display Folder name in Thread empty view

### PR DESCRIPTION
**Pull request by KevinBoulongne**
Depends on #1578

### CODE REVIEW

File: app/src/main/res/layout/fragment_thread.xml [hunk 1]
```diff
@@ -138,6 +138,18 @@
         android:layout_width="match_parent"
         android:layout_height="match_parent"
         android:visibility="visible"
-        tools:visibility="invisible" />
+        tools:visibility="invisible">
+
+        <TextView
```
Comment (FabianDevel) [inline app/src/main/res/layout/fragment_thread.xml:143]: Use a frame layout as root, and delete the emptyView's constraintLayout

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 2]
```diff
@@ -50,13 +52,31 @@ open class TwoPaneFragment : Fragment() {
...
 
+    private fun observeFolderName() = with(mainViewModel) {
+
+        Utils.waitInitMediator(currentFolder, isInSearch).observe(viewLifecycleOwner) { (folder, isInSearch) ->
+
+            val displayedFolder = (if (isInSearch) {
+                Folder().apply { name = getString(R.string.searchFolderName) }
+            } else {
+                folder
+            }) ?: return@observe
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:71]: Directly return formatted name

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 3]
```diff
@@ -337,6 +338,7 @@ class ThreadListFragment : TwoPaneFragment(), SwipeRefreshLayout.OnRefreshListen
         }
 
         searchButton.setOnClickListener {
+            mainViewModel.isInSearch.value = true
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:341]: Values of `LiveDatas` shouldn't be modified directly in the `View`, here the `Fragment`.

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 4]
```diff
@@ -51,13 +53,31 @@ open class TwoPaneFragment : Fragment() {
...
+
+            rightPaneFolderName.value = getString(
+                R.string.noConversationSelected,
+                displayedFolder.getLocalizedName(requireContext()),
+            )
+        }
+    }
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:79]: You shouldn't expose `isInSearch` and instead you can have a `LiveData` returning the folder name. That way, all the logic of which name you have to display is only in the `ViewModel`.

File: app/src/main/res/layout/fragment_thread.xml [hunk 5]
```diff
@@ -133,11 +133,17 @@
...
         android:id="@+id/emptyView"
         android:layout_width="match_parent"
         android:layout_height="match_parent"
+        android:gravity="center"
         android:visibility="visible"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        tools:text="@string/noConversationSelected"
```
Comment (tevincent) [inline app/src/main/res/layout/fragment_thread.xml:146]: You use `attrs` from `ConstraintLayout` whereas you're in a `FrameLayout` ?

Comment (KevinBoulongne) [reply app/src/main/res/layout/fragment_thread.xml:146]: Oopsy, oversight!

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 6]
```diff
@@ -51,13 +53,31 @@ open class TwoPaneFragment : Fragment() {
...
+
+            val displayedFolder = (if (isInSearch) {
+                Folder().apply { name = getString(R.string.searchFolderName) }
+            } else {
+                folder
+            }) ?: return@observe
+
+            rightPaneFolderName.value = getString(
+                R.string.noConversationSelected,
+                displayedFolder.getLocalizedName(requireContext()),
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:76]: you can use the viewModel's Context here as you're in the with

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 7]
```diff
@@ -337,6 +338,7 @@ class ThreadListFragment : TwoPaneFragment(), SwipeRefreshLayout.OnRefreshListen
...
             safeNavigate(
                 ThreadListFragmentDirections.actionThreadListFragmentToSearchFragment(
                     dummyFolderId = mainViewModel.currentFolderId ?: "eJzz9HPyjwAABGYBgQ--", // Hardcoded INBOX folder
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:344]: What about putting this inbox folder ID in the companion of the FolderController, along with the searchFolderId ?

**Review by FabianDevel — CHANGES_REQUESTED** [review]

**Review by FabianDevel — APPROVED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 8]
```diff
@@ -50,13 +53,28 @@ abstract class TwoPaneFragment : Fragment() {
...
+
+        Utils.waitInitMediator(currentFolder, isInSearch).observe(viewLifecycleOwner) { (folder, isInSearch) ->
+
+            val displayedFolder = (if (isInSearch) {
+                Folder().apply { name = getString(R.string.searchFolderName) }
+            } else {
+                folder
+            }) ?: return@observe
+
+            rightPaneFolderName.value = getString(R.string.noConversationSelected, displayedFolder.getLocalizedName(context))
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:74]: What if we create a function inside TwoPaneFragment like `onFolderNameChanged` or `updateSecondPaneFolderName` or whatever, and we call it here and the parent overrides it

**Review by LunarX — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 9]
```diff
@@ -50,13 +52,28 @@ abstract class TwoPaneFragment : Fragment() {
...
 
     private fun setupSlidingPane() {
         slidingPaneLayout.lockMode = SlidingPaneLayout.LOCK_MODE_LOCKED
     }
 
+    private fun observeFolderName() = with(mainViewModel) {
+
+        currentFolder.observe(viewLifecycleOwner) { folder ->
+
+            val displayedFolder = (if (this@TwoPaneFragment is SearchFragment) {
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:67]: If you find an elegant way to to reduce coupling for `this@TwoPaneFragment is SearchFragment` so TwoPaneFragment doesn't need to know about the existence of its possible children, please do

**Review by LunarX — CHANGES_REQUESTED** [review]

**Pull request merged**